docs/madr/decisions/016-policy-array-merging.md

```diff
@@ -0,0 +1,133 @@
+# Policy array merging
+
+* Status: accepted
+
+## Context and Problem Statement
+
+[New policies](005-policy-matching.md) support merging of multiple configuration.
+While merging regular fields is pretty straightforward, it is quite difficult to pick a strategy for merging list.
+Our current strategy is to replace list from the most specific policy.
+
+For example
+```
+type: Policy
+targetRef:
+  kind: Mesh
+defaults:
+  elements:
+  - a
+```
+and
+```
+type: Policy
+targetRef:
+  kind: MeshService
+  name: backend
+defaults:
+  elements:
+  - b
+```
+would result in
+```
+defaults:
+  elements:
+  - b
+```
+
+But for some cases we would like to append list and end up with
+```
+defaults:
+  elements:
+  - a
+  - b
+```
+
+Example use cases are:
+* Composing ProxyTemplate modifications.
+* Composing policies in Kong Mesh's OPA Policy
+* Potentially composing backends in stuff like MeshTrafficLog
+
+## Considered Options
+
+* Name convention
+* Field annotation
+* Merge keyword
+
+## Pros and Cons of the Options
+
+### Name convention
+
+We could have a name convention that indicates that given array is appendable not replaceable when we merge.
+
+```
+defaults:
+  composableElements:
+  - a
+```
+
+So every time there is the field with value array that starts with `composable` we append.
+
+Other potential prefixes:
+* appendable
+* mergeable
+* combined
+* joined
+
+To implement this, we would need to replace [jsonpatch](https://github.com/kumahq/kuma/blob/master/pkg/core/xds/rules.go#L207) with our reflection-based solution.
+
+#### Advantages
+* Explicit for users
+* Replacing jsonpatch will be more efficient, because we don't need to marshal to json just for the sake of merging
+
+#### Disadvantages
+* Longer field names
+* Reserving the prefix as a field name (although it's very unlikely to see this as a prefix of any field)
+* We need to roll out custom implementation of merging.
+
```

Comment:
I would also add that maintaining our own reflection-based solution introduces maintainability overhead and working with golang's reflections may be very tricky, so it would have to be thoroughly tested I think

Reply:
added maintainability part